pkg/cmd/server/origin/audit.go

```diff
@@ -0,0 +1,173 @@
+package origin
+
+import (
+	"bufio"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pborman/uuid"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	apiresponsewriters "k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const AuditTriggered = "audit-triggered"
+
+var _ http.ResponseWriter = &auditResponseWriter{}
+
+// auditResponseWriter is responsible for serving as a fallback audit, responsible
+// for logging failed authn events.
+type auditResponseWriter struct {
+	http.ResponseWriter
+	contextMapper apirequest.RequestContextMapper
+	req           *http.Request
+	out           io.Writer
+}
+
+func (a *auditResponseWriter) WriteHeader(code int) {
+	ctx, ok := a.contextMapper.Get(a.req)
+	if !ok {
+		apiresponsewriters.InternalError(a.ResponseWriter, a.req, errors.New("no context found for request"))
+		return
+	}
+	// if the original audit handler triggered there's no need to do anything
+	triggeredValue := ctx.Value(AuditTriggered)
+	if triggered, ok := triggeredValue.(bool); ok && triggered {
+		a.ResponseWriter.WriteHeader(code)
+		return
+	}
+	id := uuid.NewRandom().String()
+	line := fmt.Sprintf("%s AUDIT: id=%q ip=%q method=%q user=%q uri=%q\n",
```

Comment:
presenting the results of getUsername() as `user=...` seems wrong... it can attribute a request to a specific user pre-authentication, when it should be clearer that this is information the user presented that has not been verified in any way